packages/grid/x-data-grid-pro/src/components/GridRowReorderCell.tsx

```diff
@@ -0,0 +1,72 @@
+import * as React from 'react';
+import {
+  gridClasses,
+  GridEvents,
+  GridRenderCellParams,
+  GridRowEventLookup,
+  gridRowTreeDepthSelector,
+  gridSortModelSelector,
+  useGridApiContext,
+  useGridRootProps,
+  useGridSelector,
+} from '@mui/x-data-grid';
+
+const GridRowReorderCell = (params: GridRenderCellParams) => {
+  const apiRef = useGridApiContext();
+  const rootProps = useGridRootProps();
+  const sortModel = useGridSelector(apiRef, gridSortModelSelector);
+  const treeDepth = useGridSelector(apiRef, gridRowTreeDepthSelector);
+
+  // TODO: remove sortModel and treeDepth checks once row reorder is compatible
+  const isDraggable = !!rootProps.enableRowReorder && !sortModel.length && treeDepth === 1;
+
+  const publish = React.useCallback(
+    (
+        eventName: keyof GridRowEventLookup,
+        propHandler?: React.MouseEventHandler<HTMLDivElement> | undefined,
+      ): React.MouseEventHandler<HTMLDivElement> =>
+      (event) => {
+        // Ignore portal
+        // The target is not an element when triggered by a Select inside the cell
+        // See https://github.com/mui/material-ui/issues/10534
+        if (
+          (event.target as any).nodeType === 1 &&
+          !event.currentTarget.contains(event.target as Element)
+        ) {
+          return;
+        }
+
+        // The row might have been deleted
+        if (!apiRef.current.getRow(params.row.id)) {
+          return;
+        }
+
+        apiRef.current.publishEvent(eventName, apiRef.current.getRowParams(params.row.id), event);
+
+        if (propHandler) {
+          propHandler(event);
+        }
+      },
+    [apiRef, params.row.id],
+  );
+
+  const draggableEventHandlers = {
+    onDragStart: publish(GridEvents.rowDragStart),
+    onDragOver: publish(GridEvents.rowDragOver),
+    onDragEnd: publish(GridEvents.rowDragEnd),
+  };
+
+  return (
+    <div className={gridClasses.actionsCell} draggable={isDraggable} {...draggableEventHandlers}>
+      <rootProps.components.RowReorderIcon />
+
+      <div className={gridClasses.rowDraggableContainer}>{params.row.id}</div>
```

Comment:
`row.id` is the `id` attribute of the row. If the user has `getRowId` this won't work. You can use the `id` prop.

How could we allow users to customize the content shown during dragging? The ID is not useful.

AG Grid requires one column to have the `rowDrag` flag so they know which content to render: https://ag-grid.com/javascript-data-grid/row-dragging/#enabling-row-dragging

![image](https://user-images.githubusercontent.com/42154031/160430518-62debc58-8dc8-4f78-b9f2-e2d9831980a2.png)

One way is to render the `id` as default but encourage users to customize it by manually adding a column with the `__reorder__` field. We use a similar approach in the master/detail to allow to change the position of the toggle column: https://mui.com/components/data-grid/group-pivot/#customizing-the-detail-panel-toggle

Reply:
I think that would be ideal. To let the user provide a value for that column but not render it. Also, the `valueFormatter`will work in that case so it should be customizable enough. I'll update the implementation.

Reply:
I've updated it so that if you pass a value to the `__reorder__` filed or a row it will use that. If no value is passed it will default to using the `id`. I think this + the additional classes on both the cell and inner container will make it customizable enough to start using it as you can also provide a React element to be a field value.

Reply:
(TBH, when I tried it I expected it to drag the entire row, and make a gap for the dragged row to be dropped into, according to the mouse position, rather than the current approach of having a proxy for the row, and the row rearranging before it's been released.)

Reply:
It works similarly to the column reorder but I agree the UX can be improved. Regarding the value displayed while dragging - it can be the first column, which won't be a problem to be changed but it creates a problem when you reorder columns (or pin a column for that matter).

For tagging the entire row - I explored this as well but if there are a lot of columns it just looks ugly (virtualized columns are not rendered).

Reply:
> if there are a lot of columns it just looks ugly

I noticed from the benchmark examples that only one attempts to do this, and even then is just showing the data, not the row, so the approach of using a single cell makes sense.

For the default, whichever is logically first whether pinned or dragged there would be fine. And to override, allowing the developer to specify a column by name rather than having to provide a value for each row would be far preferable, although the latter option could still be retained.

Regarding the UX, the other issue I noticed is with the row highlight. If it's a selected row that's being dragged, the blue highlight follows the row being moved, but if a row is grey (mouse-over?), it doesn't, but rather moves to seemingly random rows if you scroll past the currently displayed rows. You can see it in the recording used in the Twitter post. If needed we can jump on a screens-share so I can demonstrate.

For the visual appearance of the table when dragging, when I looked at the benchmarks, I noticed that AG Grid and one other implementation use a transition that makes it very obvious what's happening, but a common pattern is to show a line at the insertion point. Not a huge fan of that, but it's clearer than the current implementation. Having the moved row clearly indicated (as it is when selected before being moved) works too. Perhaps highlight the row being moved in blue, even if it isn't selected, then remove the highlight when the moved row is "dropped"?

I wasn't too sure about the green plus icon on the cursor, since nothing is being added (appreciate this was copied from column reordering). Perhaps "grab" and grabbing" would be better?  (Not 100% sure about that either...)

And finally, the Material DragIndicator icon is quite "heavy". Perhaps it could be rendered in a lighter shade, and/or swapped with DragHandle?